teuthology/task/ship_utilities.py

```diff
@@ -0,0 +1,33 @@
+"""
+Exececute custom commands
+"""
+import contextlib
+import logging
+
+from teuthology.task.install.util import ship_utilities
+from teuthology import contextutil
+
+log = logging.getLogger(__name__)
+
+@contextlib.contextmanager
+def task(ctx, config=None):
+    """
+    Ship the utilities
+
+    ship_utilities_early: True
+    tasks:
+    - install:
+        project: ceph
+        branch: bar
+
+    The 'ship_utilities_early: True' will make the 'install' tasks skip
```

Comment:
Instead if you want a separate tasks for ship_utilities, I suppose you to use `ship_utilities` instead of `ship_utilities_early` and add to the tasks list along with other internal tasks. Taking into account that we do not do shipping by default, the default value for ship_utilities should be False. 

But first of all what is the goal of this change, what do we want to have eventually? Like install utils before the `install` task or make it possible to ship_utils without running `install` task?

Reply:
@lxbsz If I am understanding the issue correctly, each `install:` task will run the `ship_utilities` task and therefore uninstall the utilities when each individual `install` task is unwound (even if the utilities are still needed by the next `install` task's unwind handler). In that case, could we not somehow just only apply `ship_utilities` to the first `install` task by storing some state in the shared context programatically?

Reply:
> @lxbsz If I am understanding the issue correctly, each `install:` task will run the `ship_utilities` task and therefore uninstall the utilities when each individual `install` task is unwound (even if the utilities are still needed by the next `install` task's unwind handler). 

Yeah, it is. The `install:` task will ship the utilities implicitly as default. 

While the `ceph:` task is unwound, the ceph will use the `adjust-limit` utility, so this is why the `ceph:` task must be after all the `install:` tasks in the qa suite's `.yaml`, then `ceph:` task will be unwound in frist place in reverse order:
```
tasks:
- install:
- ceph:
```

Or we will hit:

```
2020-12-13T04:11:21.429 DEBUG:teuthology.run_tasks:Unwinding manager ceph
2020-12-13T04:11:21.441 INFO:tasks.ceph.ceph_manager.ceph:Canceling any pending splits or merges...
2020-12-13T04:11:21.449 DEBUG:teuthology.orchestra.run.smithi066:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph osd dump --format=json
2020-12-13T04:11:21.493 INFO:teuthology.orchestra.run.smithi066.stderr:sudo: adjust-ulimits: command not found
2020-12-13T04:11:21.494 DEBUG:teuthology.orchestra.run:got remote process result: 1
2020-12-13T04:11:21.495 ERROR:teuthology.contextutil:Saw exception from nested tasks
```

But when installing the ceph-iscsi rpm package, it will depend on the ceph packages, that means in this case the tasks order must be:

```
tasks:
- ceph:
- install:
    packages:
      rpm:
      - ceph-iscsi
    shaman:
      force_noarch: True
    project: ceph-iscsi
    sha1: latest
```
Or we will hit:

```
2020-12-13T10:03:07.260 INFO:teuthology.orchestra.run.smithi189.stderr:Error:
2020-12-13T10:03:07.261 INFO:teuthology.orchestra.run.smithi189.stderr: Problem: conflicting requests
2020-12-13T10:03:07.261 INFO:teuthology.orchestra.run.smithi189.stderr:  - nothing provides python3-rbd >= 10.2.2 needed by ceph-iscsi-3.4-47.gf79bf8c.el8.noarch
2020-12-13T10:03:07.261 INFO:teuthology.orchestra.run.smithi189.stderr:  - nothing provides python3-rados >= 10.2.2 needed by ceph-iscsi-3.4-47.gf79bf8c.el8.noarch
2020-12-13T10:03:07.261 INFO:teuthology.orchestra.run.smithi189.stderr:  - nothing provides ceph-common >= 10.2.2 needed by ceph-iscsi-3.4-47.gf79bf8c.el8.noarch
2020-12-13T10:03:07.262 INFO:teuthology.orchestra.run.smithi189.stdout:(try to add '--skip-broken' to skip uninstallable packages or '--nobest' to use not only best candidate packages)
2020-12-13T10:03:07.304 DEBUG:teuthology.orchestra.run:got remote process result: 1
2020-12-13T10:03:07.305 ERROR:teuthology.contextutil:Saw exception from nested tasks
```

This two cases above are confilicting. And this confilicting is a potential bug for the current code, to fix it we must keep the utilities shipped until the last user is unwound.

> In that case, could we not somehow just only apply `ship_utilities` to the first `install` task by storing some state in the shared context programatically?

I am afraid this won't work, because in some tasks will make the `ship_utilities` as a nested task.

> Instead if you want a separate tasks for ship_utilities, I suppose you to use `ship_utilities` instead of `ship_utilities_early` and add to the tasks list along with other internal tasks. Taking into account that we do not do shipping by default, the default value for ship_utilities should be False.
> 

Yeah, I think this sounds reasonable, but we need a small improvement, that's when iterating and running the task list one by one, we will only ship the utilities for the first task whose `ship_utilities` is `True`, for all the others just ignore the option.

> But first of all what is the goal of this change, what do we want to have eventually? Like install utils before the `install` task or make it possible to ship_utils without running `install` task?

As described above, just want to make sure that when the tasks are unwound the utilites will be kept shipped until the last user is unwounded.

Reply:
> > In that case, could we not somehow just only apply `ship_utilities` to the first `install` task by storing some state in the shared context programatically?
> 
> I am afraid this won't work, because in some tasks will make the `ship_utilities` as a nested task.

Ok, can you help me understand this better? If you have a helper that adds `ship_utilities` as a nested task only if it hasn't already been added (as denoted by a state variable in the context), the first `install` task to be executed should therefore install it and also uninstall it on unwind, right?

Reply:
In ceph code, such as:

```
qa/tasks/ceph_deploy.py:728:            lambda: install_fn.ship_utilities(ctx=ctx, config=None),
qa/tasks/ceph_deploy.py:912:        lambda: install_fn.ship_utilities(ctx=ctx, config=None),
```
Yeah, anyway we need to change this code too.

Reply:
Yup -- or make it so `ship_utilities` is the wrapper and becomes a no-op for setup and unwind if another task content was already run.

Reply:
> Yeah, I think this sounds reasonable, but we need a small improvement, that's when iterating and running the task list one by one, we will only ship the utilities for the first task whose `ship_utilities` is `True`, for all the others just ignore the option.

@dillaman Does the approach mentioned above work for you ?

Reply:
If you can do it programmatically, why would we want a new variable?

Reply:
Not only the `install:` tasks will ship the utilities, but also for some other non `install:` tasks, like the `ceph:`, etc. Shouldn't we need one way to know who need them ?

Reply:
We need to make the imlicit nested `install.ship_utilities` to be explicit in the `yaml`.

Reply:
Okay, maybe I misunderstand it, you mean do it in the `ship_utilies()` instead, right ?

Reply:
Yes. `ship_utilities` is already decorated as a context manger, so the first instance to run should test to see if it's the first. If yes, proceed as normal. Otherwise, no-op itself with just a `yield` that doesn't install nor cleanup anything.

Reply:
Sounds cool, will try it. Thanks.